doc/herdstat.1.in
.\" $Id$
.TH "@PACKAGE@" 1 "Apr 2005" "@VERSION@"
.SH NAME
\fB@PACKAGE@\fR \- Display package/developer information for Gentoo herds.
.SH SYNOPSIS
.B "@PACKAGE@ "
[ \fIoptions\fR ] [ \fI args \fR ]
.SH DESCRIPTION
\fB@PACKAGE@\fR is a utility that parses Gentoo's herds.xml and displays developer
and/or package statistics.
.SH OPTIONS
.TP
.B "\-h, \-\-help"
Display usage information.
.TP
.B "\-V, \-\-version"
Display version information.
.TP
.B "\-n, \-\-nocolor"
Don't display colored output.
.TP
.B "\-v, \-\-verbose"
Display verbose output.
.TP
.B "\-q, \-\-quiet"
Don't display labels and fancy colors.  Use this option to pipe @PACKAGE@
output to other programs.
.TP
.B "\-D, \-\-debug"
Display debugging messages.
.TP
.B "\-t, \-\-timer"
Display elapsed time of XML parsing.  Automatically enabled if --debug
is specified.
.TP
.B "\-c, \-\-count"
Display the number of items instead of the items themselves.
.TP
.B "\-\-with-herd \fI<herd>\fR"
When used in conjunction with --package --dev, only display those packages
belonging to the specified herd.
.TP
.B "\-\-no-herd"
Shorthand for --with-herd=no-herd.
.TP
.B "\-p, \-\-package"
Display package information for the specified herd(s) or developer(s).
.TP
.B "\-d, \-\-dev"
Display information for the specified developer(s).  If --package is specified,
displays package information. Otherwise, displays herd information.
.TP
.B "\-m, \-\-metadata"
Display information for the specified package(s)/categories. A package name
can be the name only (no category) only if there's not another package with the
same name.  In that case you must specify the category.  On a small note, this
option does not recognize most boolean flags (verbose, quiet, etc).
.TP
.B "\-w, \-\-which"
Display full path to ebuild for the specified package(s).
.TP
.B "\-H, \-\-herdsxml \fI<file>\fR"
Specify the location of herds.xml.  See the \fBENVIRONMENT\fR section
for an alternative way of setting the herds.xml location.
.TP
.B "\-o, \-\-outfile \fI<file>\fR"
Send output to the specified file instead of stdout.  \fI<file>\fR can also be
either stdout (default if -o is not specified) or stderr. This option implies --quiet.
.TP
.B "\-f, \-\-fetch"
Force a fetch of herds.xml.  This is only valid if a) you didn't set a herds.xml
location with -H or $HERDS, or b) the location you did set is a URL supported by
wget.  If no non-option arguments are specified, @PACKAGE@ will simply fetch
herds.xml and exit.
.SH "NON-OPTION ARGUMENTS"
Unless you specify --version or --help, at least one non-option argument is
required.  The non-option arguments depend on the action:
.TP
.B "default action" "\fI< all > [ herd(s) ]\fR"
Display developer information for the specified herd(s). If 'all' is specified,
all herds will be displayed.
.TP
.B "\-d, \-\-dev" "\fI< all > [ dev(s) ]\fR"
Display herd information for the specified developer(s). If 'all' is specified,
all developers will be displayed.
.TP
.B "\-p, \-\-package" "\fI[ herd(s)/developer(s) ]\fR"
Display package information for the specified herd(s) or developer(s).
.TP
.B "\-m, \-\-metadata" "\fI[ package(s)/categories ]\fR"
Display metadata information for the specified package(s).
.TP
If no options (non-option or not) are specified, @PACKAGE@ will display a summary of herds.xml statistics.  If --dev and --package are both specified, @PACKAGE@ will display all packages maintained by the specified developer.
.SH EXAMPLES
.IP "\fB$ @PACKAGE@ netmon x11\fR"
Display all developers in the netmon and x11 herds.
.IP "\fB$ @PACKAGE@ -m app-vim\fR"
Display metadata information for the app-vim category.
.IP "\fB$ @PACKAGE@ -m rkhunter\fR"
Display metadata information for the rkhunter package.
.IP "\fB$ @PACKAGE@ -p netmon\fR"
Display all packages in the netmon herd.
.IP "\fB$ @PACKAGE@ -d ka0ttic\fR"
Display all herds that the developer ka0ttic is in.
.IP "\fB$ @PACKAGE@ -dp ka0ttic\fR"
Display all packages ka0ttic maintains.
.IP "\fB$ @PACKAGE@ -dp ka0ttic --with-herd cron\fR"
Display all packages ka0ttic maintains that belong to the cron herd.
.IP "\fB$ @PACKAGE@ -pc base-system\fR"
Display the number of packages in the base-system herd.
.IP "\fB$ @PACKAGE@ -dpc ka0ttic\fR"
Display the number of packages maintained by ka0ttic.
.IP "\fB$ @PACKAGE@ -o ~/x11-devs.txt x11\fR"
Send the output of '@PACKAGE@ x11' to ~/x11-devs.txt.
.IP "\fB$ @PACKAGE@ all\fR"
Display all herds.
.IP "\fB$ for x in $(@PACKAGE@ -pq cron); do ... ; done\fR" 
Perform an action on every package in the cron herd.
.IP "\fB$ @PACKAGE@ $(@PACKAGE@ -dq vapier)\fR"
Display the default herd output for all herds that the developer vapier is in.
.IP "\fB$ @PACKAGE@ -pq $(@PACKAGE@ -dq ka0ttic)\fR"
Display every package that belongs to a herd that ka0ttic is a member of.
.SH ENVIRONMENT
.TP
.B HERDS
Location of herds.xml.
.TP
.B PORTDIR
Location of your portage tree.
.TP
.B ECHANGELOG_USER
If set (correctly), your developer nick will be highlighted if it occurs
in any of the output.
.SH AUTHORS
Aaron Walker <ka0ttic@gentoo.org>
.SH "REPORTING BUGS"
Please report bugs via http://bugs.gentoo.org/
.\" vim: set ft=nroff :


src/herdstat.cc
/*
 * herdstat -- src/herdstat.cc
 * $Id: herdstat.cc 237 2005-04-19 15:42:53Z ka0ttic $
 * Copyright (c) 2005 Aaron Walker <ka0ttic at gentoo.org>
 *
 * This file is part of herdstat.
 *
 * herdstat is free software; you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later
 * version.
 *
 * herdstat is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along with
 * herdstat; if not, write to the Free Software Foundation, Inc., 59 Temple
 * Place, Suite 325, Boston, MA  02111-1257  USA
 */

#ifdef HAVE_CONFIG_H
# include "config.h"
#endif

#include <iostream>
#include <locale>
#include <map>
#include <vector>
#include <algorithm>
#include <memory>
#include <cstring>
#include <cstdio>
#include <cstdlib>
#include <cstdarg>
#include <cerrno>
#include <ctime>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>

#ifdef HAVE_GETOPT_H
# include <getopt.h>
#endif

#include "common.hh"
#include "herds_xml.hh"
#include "formatter.hh"
#include "xmlparser.hh"
#include "action_herd_handler.hh"
#include "action_pkg_handler.hh"
#include "action_dev_handler.hh"
#include "action_meta_handler.hh"
#include "action_stats_handler.hh"
#include "action_which_handler.hh"
#include "action_versions_handler.hh"

static const char *short_opts = "H:o:hVvDdtpqfcnmw";

#ifdef HAVE_GETOPT_LONG
static struct option long_opts[] =
{
    {"version",	    no_argument,	0,  'V'},
    {"help",	    no_argument,	0,  'h'},
    {"verbose",	    no_argument,	0,  'v'},
    {"quiet",	    no_argument,	0,  'q'},
    {"debug",	    no_argument,	0,  'D'},
    {"count",	    no_argument,	0,  'c'},
    {"nocolor",	    no_argument,	0,  'n'},
    {"no-herd",	    no_argument,	0,  '\n'},
    {"with-herd",   required_argument,	0,  '\v'},
    /* force a fetch of herds.xml */
    {"fetch",	    no_argument,	0,  'f'},
    /* time how long it takes for XML parsing */
    {"timer",	    no_argument,	0,  't'},
    /* instead of displaying devs for a herd, display herds for a dev */
    {"dev",	    no_argument,	0,  'd'},
    /* specify the location of a local herds.xml */
    {"herdsxml",    required_argument,	0,  'H'},
    /* show package stats for the specified herds */
    {"package",	    no_argument,	0,  'p'},
    /* display package metadata information */
    {"metadata",    no_argument,	0,  'm'},
    {"which",	    no_argument,	0,  'w'},
    {"versions",    no_argument,	0,  '\b'},
    /* specify a file to write the output to */
    {"outfile",	    required_argument,	0,  'o'},
    { 0, 0, 0, 0 }
};
#endif /* HAVE_GETOPT_LONG */

void
version()
{
    std::cout << PACKAGE << "-" << VERSION << std::endl;
}

void
usage()
{
    std::cerr
	<< "usage: " << PACKAGE << " [options] [args]" << std::endl
	<< "Use --help to see more detailed usage information." << std::endl;
}

void
help()
{
    std::cout
	<< "usage: " << PACKAGE << " [options] [args]" << std::endl

#ifdef HAVE_GETOPT_LONG
	<< " -h, --help             Display this help message." << std::endl
	<< " -V, --version          Display version information." << std::endl
	<< std::endl
	<< "Where [options] can be any of the following:" << std::endl
	<< " -p, --package          Look up packages by herd." << std::endl
	<< " -d, --dev              Look up herds by developer." << std::endl
	<< " -m, --metadata         Look up metadata by package/category." << std::endl
	<< " -w, --which            Look up full path to ebuild for specified packages." << std::endl
	<< "     --versions	    Look up versions of specified packages." << std::endl
	<< "     --with-herd <herd> When used in conjunction with --package and --dev," << std::endl
	<< "                        display all packages that belong to the specified herd." << std::endl
	<< "     --no-herd          Shorthand for --with-herd=no-herd" << std::endl
	<< " -H, --herdsxml <file>  Specify location of herds.xml." << std::endl
	<< " -o, --outfile  <file>  Send output to the specified file" << std::endl
	<< "                        instead of stdout." << std::endl
	<< " -f, --fetch            Force a fetch of herds.xml." << std::endl
	<< " -v, --verbose          Display verbose output." << std::endl
	<< " -q, --quiet            Don't display labels and fancy colors. Use this"
	<< std::endl
	<< "                        option to pipe herdstat output to other programs"
	<< std::endl
	<< " -D, --debug            Display debugging messages." << std::endl
	<< " -t, --timer            Display elapsed time of XML parsing." << std::endl
	<< " -c, --count            Display the number of items instead of the" << std::endl
	<< "                        items themself." << std::endl
	<< " -n, --nocolor          Don't display colored output." << std::endl
	<< std::endl
	<< "Where [args] depends on the specified action:" << std::endl
	<< " default action         1 or more herds." << std::endl
	<< " -p, --package          1 or more herds." << std::endl
	<< " -d, --dev              1 or more developers." << std::endl
	<< " -m, --metadata         1 or more categories/packages." << std::endl
	<< " -w, --which            1 or more packages." << std::endl
	<< std::endl
	<< "Both the default action and the --dev action support an 'all' target" << std::endl
	<< "that show all of the devs or herds.  If both --dev and --package are" << std::endl
	<< "specified, " << PACKAGE << " will display all packages maintained by" << std::endl
	<< "the specified developer." << std::endl
#else

	<< " -h              Display this help message." << std::endl
	<< " -V              Display version information." << std::endl
	<< std::endl
	<< "Where [options] can be any of the following:" << std::endl
	<< " -p              Look up packages by herd." << std::endl
	<< " -d              Look up herds by developer." << std::endl
	<< " -m              Look up metadata by package/category." << std::endl
	<< " -w              Look up full path to ebuild for specified packages." << std::endl
	<< " -H <file>       Specify location of herds.xml." << std::endl
	<< " -o <file>       Send output to the specified file" << std::endl
	<< "                 instead of stdout." << std::endl
	<< " -f              Force a fetch of herds.xml." << std::endl
	<< " -v              Display verbose output." << std::endl
	<< " -q              Don't display labels and fancy colors. Use this"
	<< std::endl
	<< "                 option to pipe herdstat output to other programs"
	<< std::endl
	<< " -D              Display debugging messages." << std::endl
	<< " -t              Display elapsed time of XML parsing." << std::endl
	<< " -c              Display the number of items instead of the" << std::endl
	<< "                 items themself." << std::endl
	<< " -n              Don't display colored output." << std::endl
	<< std::endl
	<< "Where [args] depends on the specified action:" << std::endl
	<< " default action  1 or more herds." << std::endl
	<< " -p              1 or more herds." << std::endl
	<< " -d              1 or more developers." << std::endl
	<< " -m              1 or more categories/packages." << std::endl
	<< " -w              1 or more packages." << std::endl
	<< std::endl
	<< "Both the default action and the -d action support an 'all' target" << std::endl
	<< "that show all of the devs or herds.  If both -d and -p are specified," << std::endl
	<< PACKAGE << " will display all packages maintained by the specified" << std::endl
	<< "developer." << std::endl

#endif /* HAVE_GETOPT_LONG */

	<< std::endl
	<< PACKAGE << " respects the HERDS environment variable." << std::endl
	<< "Set it in your shell rcfile to permanently set the location of "
	<< "your herds.xml." << std::endl;
}

int
handle_opts(int argc, char **argv, std::vector<std::string> *args)
{
    int key, opt_index = 0;

    while (true)
    {
#ifdef HAVE_GETOPT_LONG
	key = getopt_long(argc, argv, short_opts, long_opts, &opt_index);
#else
	key = getopt(argc, argv, short_opts);
#endif /* HAVE_GETOPT_LON)G */

	if (key == -1)
	    break;

	switch (key)
	{
	    /* --dev */
	    case 'd':
		if (optget("action", options_action_T) != action_unspecified and
		    optget("action", options_action_T) != action_pkg)
		    throw args_one_action_only_E();
		if (optget("action", options_action_T) == action_pkg)
		    optset("dev", bool, true);
		else
		    optset("action", options_action_T, action_dev);
		break;
	    /* --package */
	    case 'p':
		if (optget("action", options_action_T) != action_unspecified and
		    optget("action", options_action_T) != action_dev)
		    throw args_one_action_only_E();
		if (optget("action", options_action_T) == action_dev)
		    optset("dev", bool, true);
		optset("action", options_action_T, action_pkg);
		break;
	    /* --metadata */
	    case 'm':
		if (optget("action", options_action_T) != action_unspecified)
		    throw args_one_action_only_E();
		optset("action", options_action_T, action_meta);
		break;
	    /* --which */
	    case 'w':
		if (optget("action", options_action_T) != action_unspecified)
		    throw args_one_action_only_E();
		optset("action", options_action_T, action_which);
		break;
	    /* --versions */
	    case '\b':
		if (optget("action", options_action_T) != action_unspecified)
		    throw args_one_action_only_E();
		optset("action", options_action_T, action_versions);
		break;
	    /* --outfile */
	    case 'o':
		if (strcmp(optarg, "stdout") != 0)
		{
		    if (strcmp(optarg, "stderr") == 0)
			optset("outstream", std::ostream *, &std::cerr);
		    optset("outfile", std::string, optarg);
		    optset("quiet", bool, true);
		    optset("timer", bool, false);
		}
		break;
	    /* --no-herd */
	    case '\n':
		optset("with-herd", std::string, "no-herd");
		break;
	    /* --with-herd */
	    case '\v':
		optset("with-herd", std::string, optarg);
		break;
	    /* --fetch */
	    case 'f':
		optset("fetch", bool, true);
		break;
	    /* --verbose */
	    case 'v':
		optset("verbose", bool, true);
		break;
	    /* --quiet */
	    case 'q':
		optset("quiet", bool, true);
		break;
	    /* --count */
	    case 'c':
		optset("count", bool, true);
		optset("quiet", bool, true);
		break;
	    /* --nocolor */
	    case 'n':
		optset("color", bool, false);
		break;
	    /* --herdsxml */
	    case 'H':
		optset("herds.xml", std::string, optarg);
		break;
	    /* --debug */
	    case 'D':
		optset("timer", bool, true);
		optset("debug", bool, true);
		break;
	    /* --timer */
	    case 't':
		if (optget("outfile", std::string) == "stdout")
		    optset("timer", bool, true);
		break;
	    /* --version */
	    case 'V':
		throw args_version_E();
		break;
	    /* --help */
	    case 'h':
		throw args_help_E();
		break;
	    case 0:
		throw args_usage_E();
		break;
	    default:
		throw args_E();
		break;
	}
    }

    if (optind < argc)
    {
	while (optind < argc)
	    args->push_back(argv[optind++]);
    }
    else if ((optget("action", options_action_T) != action_unspecified) and
	    (optget("action", options_action_T) != action_meta))
	throw args_usage_E();

    return 0;
}

int
main(int argc, char **argv)
{
    options_T options;
    util::color_map_T color;

    /* try to determine current columns, otherwise use default */
    optset("maxcol", std::size_t, util::getcols());

    try
    { 
	std::vector<std::string> nonopt_args;

	/* handle command line options */
	if (handle_opts(argc, argv, &nonopt_args) != 0)
	    throw args_E();

	if (nonopt_args.empty() and
	    optget("action", options_action_T) != action_meta)
	{
	    optset("action", options_action_T, action_stats);
	    optset("quiet", bool, false);
	}

	/* remove duplicates; also has the nice side advantage
	 * of sorting the output */
	std::sort(nonopt_args.begin(), nonopt_args.end());
	std::vector<std::string>::iterator pos =
	    std::unique(nonopt_args.begin(), nonopt_args.end());
	if (pos != nonopt_args.end())
	    nonopt_args.erase(pos);

	/* did the user specify the all target? */
	if (std::find(nonopt_args.begin(),
		nonopt_args.end(), "all") != nonopt_args.end())
	{
	    optset("all", bool, true);
	    nonopt_args.clear();
	    nonopt_args.push_back("all");
	}

	/* dump options */
	if (optget("debug", bool))
	    options.dump(*optget("outstream", std::ostream *));

	/* --fetch and no options: fetch herds.xml and exit */
	if (optget("fetch", bool) and nonopt_args.empty())
	{
	    if (not optget("quiet", bool))
		optset("verbose", bool, true);

	    herds_xml_T herdsxml(true);
	    return EXIT_SUCCESS;
	}

	/* setup outfile */
	std::ostream *outstream = NULL;
	if (optget("outfile", std::string) != "stdout" and
	    optget("outfile", std::string) != "stderr")
	{
	    outstream = new std::ofstream(optget("outfile", std::string).c_str());
	    if (not *outstream)
		throw bad_fileobject_E(optget("outfile", std::string));
	    optset("outstream", std::ostream *, outstream);
	}
	else
	{
	    try
	    {
		optset("locale", std::string, std::locale("").name());
	    }
	    catch (const std::runtime_error)
	    {
		std::string error("Invalid locale");
		char *result = std::getenv("LC_ALL");
		if (result)
		    error += " '" + std::string(result) + "'.";
		std::cerr << error << std::endl;
		return EXIT_FAILURE;
	    }
	}
	
	/* set locale */
	optget("outstream", std::ostream *)->imbue
	    (std::locale(optget("locale", std::string).c_str()));

	/* set common format attributes */
	formatter_T output;
	output.set_colors(optget("color", bool));
	output.set_quiet(optget("quiet", bool));
	output.set_labelcolor(color[green]);
	output.set_highlightcolor(color[yellow]);
	output.add_highlight(util::current_user());
	output.add_highlight(util::get_user_from_email(util::current_user()));

	/* set default action */
	if (optget("action", options_action_T) == action_unspecified)
	    optset("action", options_action_T, action_herd);

	std::map<options_action_T, action_handler_T * > handlers;
	handlers[action_herd]  = new action_herd_handler_T();
	handlers[action_dev]   = new action_dev_handler_T();
	handlers[action_pkg]   = new action_pkg_handler_T();
	handlers[action_meta]  = new action_meta_handler_T();
	handlers[action_stats] = new action_stats_handler_T();
	handlers[action_which] = new action_which_handler_T();
	handlers[action_versions] = new action_versions_handler_T();

	action_handler_T *action_handler =
	    handlers[optget("action", options_action_T)];
	if (action_handler)
	{
	    try
	    {
		if ((*action_handler)(nonopt_args) != EXIT_SUCCESS)
		    return EXIT_FAILURE;
	    }
	    catch (action_E)
	    {
		return EXIT_FAILURE;
	    }
	}
	else
	    throw args_unimplemented_E();

	if (outstream)
	    delete outstream;

    }
    catch (const XMLParser_E &e)
    {
	std::cerr << "Error parsing '" << e.file() << "': " << e.error()
	    << std::endl;
	return EXIT_FAILURE;
    }
    catch (const fetch_E)
    {
	return EXIT_FAILURE;
    }
    catch (const util::base_E &e)
    {
	std::cerr << e.what() << std::endl;
	return EXIT_FAILURE;
    }
    catch (const errno_error_E &e)
    {
	std::cerr << e.what() << std::endl;
	return EXIT_FAILURE;
    }
    catch (const timer_E &e)
    {
	std::cout
	    << "Took " << e.what() << "ms to parse herds.xml." << std::endl;
    }
    catch (const args_help_E)
    {
	help();
	return EXIT_SUCCESS;
    }
    catch (const args_version_E)
    {
	version();
	return EXIT_SUCCESS;
    }
    catch (const args_usage_E)
    {
	usage();
	return EXIT_FAILURE;
    }
    catch (const args_E)
    {
	usage();
	return EXIT_FAILURE;
    }
    catch (const herdstat_base_E &e)
    {
	std::cerr << "Unhandled exception: " << e.what() << std::endl;
	return EXIT_FAILURE;
    }

    return EXIT_SUCCESS;
}
